### DISCUSSION TITLE
mbstowcs/wcstombs и многопоточность в линуксе

### DISCUSSION THREAD

**Discussion by plzombie** [plzombie/djvupure#2] [General]
Смотрю я значит в очередной раз справку по mbstowcs/wcstombs

Сайт https://en.cppreference.com/w/cpp/string/multibyte/mbstowcs говорит 

> Each character is converted as if by a call to [std::mbtowc](https://en.cppreference.com/w/cpp/string/multibyte/mbtowc), except that the mbtowc conversion state is unaffected.

Но при этом

> In most implementations, this function updates a global static object of type [std::mbstate_t](https://en.cppreference.com/w/cpp/string/multibyte/mbstate_t) as it processes through the string, and cannot be called simultaneously by two threads, [std::mbsrtowcs](https://en.cppreference.com/w/cpp/string/multibyte/mbsrtowcs) should be used in such cases.

Смотрю, как реализовано в windows 10. Конкретно для mbstowcs используется оптимизация и всё сводится тупо к вызову MultibyteToWidechar из winapi (хотя для mbsrtowcs, действительно, вызывается mbrtowc в цикле) или
```
mbstate state{};
return _mbstowcs_utf8(...,&state,...);
```
. mbtowc вообще не использует mbstate,
Для wcstombs аналогично используется WidecharToMultibyte. Для wctomb аналогично никаких глобальных состояний не меняется.

Насчёт линукса. Вот здесь вот https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_09_01 функции отсутствуют
В манах для соляриса https://www.opennet.ru/man.shtml?topic=mbstowcs&category=3&russian=4 написано MT-Level | MT-Safe, тоже всё ок
Но вот здесь https://www.opennet.ru/man.shtml?topic=wcstombs&category=3&russian=5 написано, что
> The wcstombs() function need not be reentrant. A function that is not required to be reentrant is not required to be thread-safe. 

Хотя в свежей версии спеки этого уже нет https://pubs.opengroup.org/onlinepubs/9699919799/functions/wcstombs.html#

Собственно, вопрос. Так их можно использовать в многопоточных приложениях или нет?

**Comment by plzombie** [comment]
В винде так-то даже asctime и rand являются thread safe, так как все глобальные значения хранятся в thread local storage

**Comment by plzombie** [comment]
На сайте man7 тоже написано, что они thread safe https://man7.org/linux/man-pages/man3/wcstombs.3.html https://man7.org/linux/man-pages/man3/mbstowcs.3.html

@zvezdochiot Если у тебя будет время свободное, можешь пожалуйста посмотреть, что там творится в glibc?

**Reply by plzombie** [reply]
Сам посмотрел, там:
``` c
/* Copyright (C) 1991-2023 Free Software Foundation, Inc.
   This file is part of the GNU C Library.

   The GNU C Library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   The GNU C Library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with the GNU C Library; if not, see
   <https://www.gnu.org/licenses/>.  */

#include <stdlib.h>
#include <string.h>
#include <wchar.h>


/* Convert the string of multibyte characters in S to `wchar_t's in
   PWCS, writing no more than N.  Return the number written,
   or (size_t) -1 if an invalid multibyte character is encountered.  */
size_t
mbstowcs (wchar_t *pwcs, const char *s, size_t n)
{
  mbstate_t state;

  memset (&state, '\0', sizeof state);
  /* Return how many we wrote (or maybe an error).  */
  return __mbsrtowcs (pwcs, &s, n, &state);
}
```
И судя по логам, так уже 24 года

```
1999-07-28  Ulrich Drepper  <drepper@cygnus.com>

	* stdlib/mblen.c: Use static state.
	Reported by Bruno Haible <haible@ilog.fr>.

	* stdlib/mbtowc.c: Reset state for s == NULL.
	* stdlib/wctomb.c: Likewise.
	Reported by Bruno Haible <haible@ilog.fr>.

	* stdlib/mbstowcs.c: Do not use global state.
	Reported by Bruno Haible <haible@ilog.fr>.
```

А до версии 2.2 было вот такое вот:
``` c
/* Copyright (C) 1991, 1992, 1996, 1997 Free Software Foundation, Inc.
   This file is part of the GNU C Library.

   The GNU C Library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public License as
   published by the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   The GNU C Library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.

   You should have received a copy of the GNU Library General Public
   License along with the GNU C Library; see the file COPYING.LIB.  If not,
   write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
   Boston, MA 02111-1307, USA.  */

#include <stdlib.h>
#include <wchar.h>


extern mbstate_t __no_r_state;	/* Defined in mbtowc.c.  */

/* Convert the string of multibyte characters in S to `wchar_t's in
   PWCS, writing no more than N.  Return the number written,
   or (size_t) -1 if an invalid multibyte character is encountered.

   Attention: this function should NEVER be intentionally used.
   The interface is completely stupid.  The state is shared between
   all conversion functions.  You should use instead the restartable
   version `mbsrtowcs'.  */
size_t
mbstowcs (wchar_t *pwcs, const char *s, size_t n)
{
  mbstate_t save_shift = __no_r_state;
  size_t written;

  written = __mbsrtowcs (pwcs, &s, n, &__no_r_state);

  /* Restore the old shift state.  */
  __no_r_state = save_shift;

  /* Return how many we wrote (or maybe an error).  */
  return written;
}
```
Удивительно, а главное непонятно, зачем.

**Comment by zvezdochiot** [comment]
@plzombie say:
>непонятно, зачем.

Звиняюсь, но подсобить не смогу ничем. Для меня вся это многопоточность - откровенная тарабарщина. Я из-за неё на `opuscomment` запнулся и бросил идею портировать под винду. Хотя очень хотел. Такие вот дела.
